daemon/mgr/container_upgrade.go

```diff
@@ -0,0 +1,219 @@
+	if err != nil {
+		return err
+	}
+
+	var (
+		needRollback  = false
+		oldConfig     = *c.Config
+		oldHostconfig = *c.HostConfig
+	)
+
+	defer func() {
+		if !needRollback {
+			return
+		}
+
+		c.Lock()
+		// recover old container config
+		c.Config = &oldConfig
+		c.HostConfig = &oldHostconfig
+		c.Image = oldConfig.Image
+		c.Unlock()
+
+		if err := mgr.createContainerdContainer(ctx, c, "", ""); err != nil {
+			logrus.Errorf("failed to rollback upgrade action: %s", err.Error())
+			if err := mgr.markStoppedAndRelease(c, nil); err != nil {
+				logrus.Errorf("failed to mark container %s stop status: %s", c.ID, err.Error())
+			}
+		}
+	}()
+
+	// merge image config to container config
+	if err := mgr.mergeImageConfigForUpgrade(ctx, c, config); err != nil {
+		return errors.Wrap(err, "failed to upgrade container")
+	}
+
+	// if the container is running, we need first stop it.
+	if c.IsRunning() {
+		if _, err := mgr.Client.DestroyContainer(ctx, c.Key(), 10); err != nil {
+			return errors.Wrapf(err, "failed to destroy container")
+		}
+	}
+
+	// prepare new snapshot for the new container
+	oldSnapID := c.SnapshotKey()
+	newSnapID, err := mgr.prepareSnapshotForUpgrade(ctx, c.Key(), c.SnapshotKey(), config.Image)
+	if err != nil {
+		return err
+	}
+	c.SetSnapshotID(newSnapID)
+
+	// initialize container storage config before container started
+	if err := mgr.initContainerStorage(ctx, c); err != nil {
+		return errors.Wrapf(err, "failed to init container storage, id: (%s)", c.Key())
+	}
+
+	// If container is running, we also should start the container
+	// after recreate it.
+	if c.IsRunning() {
+		if err := mgr.createContainerdContainer(ctx, c, "", ""); err != nil {
+			needRollback = true
+			if err := mgr.Client.RemoveSnapshot(ctx, newSnapID); err != nil {
+				logrus.Errorf("failed to remove snapshot %s: %v", newSnapID, err)
+			}
+
+			return errors.Wrap(err, "failed to create new container")
+		}
+	}
+
+	// Upgrade success, remove snapshot of old container
+	if err := mgr.Client.RemoveSnapshot(ctx, oldSnapID); err != nil {
+		// TODO(ziren): remove old snapshot failed, may cause dirty data
+		logrus.Errorf("failed to remove snapshot %s: %v", oldSnapID, err)
+	}
+
+	// Upgrade succeeded, refresh the cache
+	mgr.cache.Put(c.ID, c)
+
+	// Works fine, store new container info to disk.
+	if err := c.Write(mgr.Store); err != nil {
+		logrus.Errorf("failed to update container %s in meta store: %v", c.ID, err)
+		return err
+	}
+
+	return nil
+}
+
+// getImageRef return the image id and name if fetch successed, otherwise return error
+func (mgr *ContainerManager) getImageRef(ctx context.Context, image string) (digest.Digest, string, error) {
+	imageID, _, primaryRef, err := mgr.ImageMgr.CheckReference(ctx, image)
+	if err != nil {
+		return "", "", err
+	}
+
+	return imageID, primaryRef.String(), nil
+}
+
+func (mgr *ContainerManager) getOCIImageConfig(ctx context.Context, image string) (ocispec.ImageConfig, error) {
```

Comment:
Please make this function owned by `ImageManager` rather than `ContainerManager`. Since this action does nothing to do with ContainerManager.